### PR TITLE
Break operand-result ties for constant operands

### PR DESCRIPTION
**Pull request by antiagainst**
We won't create such ties deliberately, but they can come from
multiple rounds of canonicalization. For example, if we have
`linalg.fill` to prepare the initial result tensor for `subtensor_insert`,
the `linalg.fill` may be folded into a constant. Then the
`subtensor_insert` op will have a tie between the initial
output tensor and its resultant tensor.

In IREE, operand-result tying indicates that we can use the same
buffer. This is unlikely to happen if the operand is constant,
though: constants are typically collected into a dense buffer
that is read only. So having such kind of ties can mess up buffer
allocation in later stages: we will try to reuse the buffer for
the constant in a read-write fashion, which is incorrect. This
becomes a problem with constant pooling after
https://github.com/google/iree/commit/4f86152fda78b7afd2351d4af5b2c85b73805dcf.

Along the way, fixed two other indexing issues related to operand-
result ties.

### CODE REVIEW

**Review by benvanik — CHANGES_REQUESTED** [review]
Great fixes! Can you please split them from the untying pattern? The untying isn't going to work in general and if we do land it as a workaround it'd need to be something that can be reverted easily.

Tied constants at the flow level are actually fine - they indicate copy-on-write behavior. When lowering into the HAL we should (but currently don't) turn it into a clone of the constant.

The bigger issue, though, is that you can't know at this level if the value is constant: a function or block argument or result of an std.select (or custom op, etc) will not match this pattern. For this we need DFA to determine when something can _possibly_ be constant and then emit the proper conditional hal IR to do the copy-on-write behavior.

This only works now because we inline everything and tensorflow has some quirks in its representation of variables and such. Other frontends won't be so simple: it's legit python to store a reference to a constant in a variable as an initializer and then overwrite it with a mutable value in the future - usage of the variable (which can then be passed through functions/returned from accessors/etc) will not be knowable at compile time and require the same conditional copy-on-write behavior as above.

Instead as a workaround for now you can (probably) extend the InsertImmutabilityPreservingStreamClones to clone not just when there are multiple users (`hasUsersInStreamAfterUpdate`) but when the input is a constant - that will add the flow.tensor.clone to model the copy-on-write behavior. Once we have a DFA pass propagating tensor usage it can be moved into that and done more generally but it should unblock you for now.

**Comment by benvanik** [comment]
(#5492 is the issue I'm using to track the DFA work, and #5472 is blocked on the same(ish) copy-on-write requiring behavior)

**Comment by antiagainst** [comment]
https://github.com/google/iree/pull/6174 superseded this.

**Pull request closed without merging**
